applications/callflow/src/cf_exe.erl

```diff
@@ -805,7 +805,52 @@ update_actions(Action, Call) ->
 %%------------------------------------------------------------------------------
 -spec spawn_cf_module(atom(), kz_json:object(), kapps_call:call(), pid()) -> kz_term:pid_ref().
 spawn_cf_module(CFModule, Data, Call, AMQPWorker) ->
-    kz_util:spawn_monitor(fun cf_module_task/4, [CFModule, Data, Call, AMQPWorker]).
+    kz_util:spawn_monitor(fun cf_maybe_blacklisted/4, [CFModule, Data, Call, AMQPWorker]).
+
+-spec cf_maybe_blacklisted(atom(), kz_json:object(), kapps_call:call(), pid()) -> any().
+cf_maybe_blacklisted(CFModule, Data, Call, AMQPWorker) ->
+    AccountBlAction = kapps_call:account_bl_action(Call),
+    case cf_try_user_blacklists(CFModule, Data, AccountBlAction, Call) of
+        #cf_bl_action{action = <<"block">>} ->
+            lager:info("action ~s execution is blocked by blacklists", [CFModule]),
+            _ = kapps_call_command:response(<<"603">>, <<"Decline">>, Call),
+            stop(Call);
+        #cf_bl_action{caller_name = 'undefined', action = ActionValue} ->
+            UpdatedCall = kapps_call:set_user_bl_action(ActionValue, Call),
+            cf_module_task(CFModule, Data, UpdatedCall, AMQPWorker);
+        #cf_bl_action{caller_name = CallerName, action = ActionValue} ->
+            UpdatedCall1 = kapps_call:set_user_bl_action(ActionValue, Call),
+            UpdatedCall2 = kapps_call:set_caller_id_name(CallerName, UpdatedCall1),
+            cf_module_task(CFModule, Data, UpdatedCall2, AMQPWorker)
+    end.
+
+-spec cf_try_user_blacklists(atom(), kz_json:object(), kz_term:api_ne_binary(), kapps_call:call()) -> cf_bl_action().
+cf_try_user_blacklists(CFModule, Data, AccountBlAction, Call) ->
+    try CFModule:bl_action(Data, Call) of
```

Comment:
If this is going to be applied to all callflow action modules, it should be an optional behaviour callback of `gen_cf_action` and the code here should check if the function is exported (using `kz_module:is_exported/3`) before calling it. The only define the function in the callflow actions that actually support it.